### PR TITLE
Handle builders in a typesafe way

### PR DESCRIPTION
**Pull request by spawnia**
Separate the handling of database query builders and the Scout query builder, since they vary wildly in feature set.

This alleviates the burden on `ArgBuilderDirectives` to add special handling for Scout builders, while centralizing the semantics of what a successful Scout search looks like.

### CODE REVIEW

**Comment by spawnia** [comment]
@olivernybroe while adding field builder support to `@like`, i ran into type issues. I want to resolve this once and for all.

File: src/Schema/Directives/WhereDirective.php [hunk 1]
```diff
@@ -34,12 +36,13 @@ public static function definition(): string
...
     /**
      * Add any "WHERE" clause to the builder.
-     *
-     * @param  \Illuminate\Database\Query\Builder|\Illuminate\Database\Eloquent\Builder  $builder
-     * @return \Illuminate\Database\Query\Builder|\Illuminate\Database\Eloquent\Builder
      */
     public function handleBuilder($builder, $value): object
     {
+        if ($builder instanceof ScoutBuilder) {
+            throw new Exception("Using {$this->name()} on queries that use a Scout search is not supported.");
```
Comment (olivernybroe) [inline src/Schema/Directives/WhereDirective.php:43]: Yeah, we definitely have to refactor this to a custom exception, so this string is not duplicated all the time 👍

File: src/Schema/Directives/WithRelationDirective.php [hunk 2]
```diff
@@ -64,9 +65,13 @@ protected function loadRelation(Model $parent, ResolveInfo $resolveInfo): Deferr
      */
     protected function decorateBuilder(ResolveInfo $resolveInfo): Closure
     {
-        return function (object $query) use ($resolveInfo): void {
+        return function (object $builder) use ($resolveInfo): void {
+            if ($builder instanceof Relation) {
```
Comment (olivernybroe) [inline src/Schema/Directives/WithRelationDirective.php:69]: Unsure what consequences this can have.

**Review by olivernybroe — CHANGES_REQUESTED** [review]
Instead of adding the if statement in all directives, what if we create a new interface with method `handleScoutBuilder`?

**Comment by olivernybroe** [comment]
Really nice. I think it would be awesome to solve this, I also found it quite annoying.

File: src/Schema/Directives/RelationDirective.php [hunk 3]
```diff
@@ -70,6 +71,10 @@ function (Model $parent, array $args, GraphQLContext $context, ResolveInfo $reso
     protected function makeBuilderDecorator(ResolveInfo $resolveInfo): Closure
     {
         return function (object $builder) use ($resolveInfo) {
+            if ($builder instanceof Relation) {
+                $builder = $builder->getQuery();
```
[the remark this answers is not in the document]
Comment (spawnia) [reply src/Schema/Directives/RelationDirective.php:75]: The tests show none, but theoretically someone might have added a typehint or code depending on a `Relation` being passed? I actually doubt it, and all that the `Relation` did was to proxy calls to `Builder´ anyway

**Comment by spawnia** [comment]
@olivernybroe this is still rough around the edges, but i found a way to handle everything related to Scout separately. This change should clarify the difference between search and other kinds of querying, for directive developers and API consumers alike.

File: src/Execution/Arguments/ArgumentSet.php [hunk 4]
```diff
@@ -141,14 +143,19 @@ function ($value) {
...
      * @param  array<string>  $scopes
      * @param  \Closure  $directiveFilter
      *
-     * @return \Illuminate\Database\Query\Builder|\Illuminate\Database\Eloquent\Builder|\Illuminate\Database\Eloquent\Model|\Illuminate\Database\Eloquent\Relations\Relation
+     * @return \Illuminate\Database\Query\Builder|\Illuminate\Database\Eloquent\Builder|\Laravel\Scout\Builder
      */
     public function enhanceBuilder(object $builder, array $scopes, Closure $directiveFilter = null): object
     {
+        $scoutEnhancer = new ScoutEnhancer($this, $builder);
+        if ($scoutEnhancer->containsSearch()) {
```
Comment (olivernybroe) [inline src/Execution/Arguments/ArgumentSet.php:155]: Naming of this method makes it look like the Scout Enhancer is the one that contains search. 
Also it is not clear that you are checking for if a directive exist. 

Consider naming it something like `isInQuery` (hopefully better name than this). But my point was to show that the naming does not need to say something that specific about the implementation.

File: src/Execution/Arguments/ArgumentSet.php [hunk 5]
```diff
@@ -141,14 +143,19 @@ function ($value) {
...
+            return $scoutEnhancer->enhance();
```
Comment (olivernybroe) [inline src/Execution/Arguments/ArgumentSet.php:156]: Maybe name it `EnhanceBuilder`

File: src/Schema/Directives/EqDirective.php [hunk 6]
```diff
@@ -22,16 +24,18 @@ public static function definition(): string
...
     public function handleBuilder($builder, $value): object
     {
         return $builder->where(
-            $this->directiveArgValue('key', $this->nodeName()),
+            $this->directiveArgValue('key') ?? $this->nodeName(),
+            $value
+        );
+    }
+
+    public function handleScoutBuilder(ScoutBuilder $builder, $value)
```
Comment (olivernybroe) [inline src/Schema/Directives/EqDirective.php:35]: Nice, this is soo clean.

File: src/Scout/ScoutEnhancer.php [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+
+class ScoutEnhancer
+{
+    /**
+     * @var ArgumentSet
+     */
+    protected $argumentSet;
+
+    /**
+     * @var object
```
Comment (olivernybroe) [inline src/Scout/ScoutEnhancer.php:22]: We should be able to change this to the Scout builder type right?

File: src/SoftDeletes/TrashedDirective.php [hunk 8]
```diff
@@ -29,28 +35,62 @@ public static function definition(): string
...
+        $model = $builder->model;
+
+        $this->assertModelUsesSoftDeletes($model);
+
+        if ($value === null) {
+            return $builder;
+        }
+
+        switch ($value) {
+            case 'with':
```
Comment (olivernybroe) [inline src/SoftDeletes/TrashedDirective.php:78]: Does without not work with scout?

**Review by olivernybroe — COMMENTED** [review]
In general I think this is a really great path and cleans things up quite a lot. 

As you said it's still a bit rough, but I think the concept is totally right. 

Mostly I think the confusing logic is inside the Scout Enhancer. 

But the directives are soo clean with this approach, love it

File: src/Scout/ScoutEnhancer.php (hunk 7 above)
Comment (spawnia) [reply src/Scout/ScoutEnhancer.php:22]: This one is actually the non-scout builder

File: src/SoftDeletes/TrashedDirective.php (hunk 8 above)
Comment (spawnia) [reply src/SoftDeletes/TrashedDirective.php:78]: Apparently not, since the builder lacks the method. I see this discovery as a big win of this PR 💪

Comment (olivernybroe) [reply src/SoftDeletes/TrashedDirective.php:78]: Yeah, this is a perfect example of how awesome this change is!

**Pull request merged**